scalableminds/webknossos
# Abort bucket pulls if a layer is reloaded

When a layer is reloaded, all buckets are "collected" which means they are deleted and reset. However, bucket requests that were still in progress were not aborted, before, which lead to `Error: Unexpected state: UNREQUESTED` errors (and potentially caused the rendering of old data if the actual data on the hard disk really changed).

Initially, I focused on aborting the actual fetch requests. However, this is fairly complex as AbortSignals cannot be transferred to web workers and in the end it didn't prevent the "UNREQUESTED" errors. There's a lot more that happens, besides the `fetch` call, in the time from when the pullqueue triggers the bucket fetching procedure until the actual array buffer is saved in the bucket. Therefore, I opted for a more general approach, allowing to abort the bucket fetching procedure at any time. I used the same `AbortController` and `AbortSignal` concept for that as it is fairly elegant. In theory, the signal could be forwarded to the actual fetch call so that it can be aborted in-flight as well, however, I'm not sure whether this limited time save is worth the substantial additional complexity.
In my tests, the layer reloading was very quick with the current approach.

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Move around and (while buckets are still loading) reload a layer using the button in the layer settings. Alternatively, reload a layer twice in quick succession.
- Open the version view and quickly switch between different volume versions.
- No errors should be shown in the console.

### Issues:
- fixes #4895 

------
